wremnants/include/muon_validation.h

```diff
@@ -56,23 +52,128 @@ class JpsiCorrectionsHelper {
     }
     
     // for central value of pt
-    float operator() (float cvh_pt, float cvh_eta, int charge) {
-        const auto &params = get_tensor(cvh_eta);
+    float operator() (float cvhPt, float cvhEta, int charge) {
+        const auto &params = get_tensor(cvhEta);
         const double A = params(0, 0);
         const double e = params(1, 0);
         const double M = params(2, 0);
-        double k = 1.0 / cvh_pt;
+        double k = 1.0 / cvhPt;
         double magnetic = 1.0 + A;
         double material = -1.0 * e * k;
         double alignment = charge * M;
-        double k_crctd = (magnetic + material) * k + alignment;
-        return (1.0 / k_crctd);
+        double kCrctd = (magnetic + material) * k + alignment;
+        return (1.0 / kCrctd);
     }
 
 private:
     std::shared_ptr<const T> correctionHist_;
+
+};
+
+template <typename T>
+class JpsiCorrectionsUncHelper {
+
+public:
+    using hist_t = T;
+    using tensor_t = typename T::storage_type::value_type::tensor_t;
+    static constexpr auto sizes = narf::tensor_traits<tensor_t>::sizes;
+    static constexpr auto nUnc = sizes[sizes.size() - 1]; // 1 for central value
+    using out_tensor_t = Eigen::TensorFixedSize<double, Eigen::Sizes<nUnc, 2>>;
+
+    JpsiCorrectionsUncHelper(T&& corrections) :
+        correctionHist_(std::make_shared<const T>(std::move(corrections))) {}
+
+    // helper for bin lookup which implements the compile-time loop over axes
+    template<typename... Xs, std::size_t... Idxs>
+    const tensor_t &get_tensor_impl(std::index_sequence<Idxs...>, const Xs&... xs) {
+        return correctionHist_->at(correctionHist_->template axis<Idxs>().index(xs)...).data();
+    }
+
+    // variadic templated bin lookup
+    template<typename... Xs>
+    const tensor_t &get_tensor(const Xs&... xs) {
+        return get_tensor_impl(std::index_sequence_for<Xs...>{}, xs...);
+    }
+
+    // for smearing weights derived from propagating uncs on A, e, M to uncs on qop
+    out_tensor_t operator() (
+        double genQop, float genPhi, int genCharge, float genEta, float genPt,
```

Comment:
Could get charge from the sign of qop and slightly reduce the number of arguments you need

Reply:
I got rid of genCharge and genPt, which are only used to calculate genQop; and like you said, we don't really need it. The calculation is only needed for RECO which is smeared GEN at the moment.